src/server/frontend_xwayland/xwayland_connector.cpp

```diff
@@ -126,9 +115,43 @@ auto mf::XWaylandConnector::socket_name() const -> optional_value<std::string>
     }
 }
 
+void mf::XWaylandConnector::maybe_create_spawner(std::unique_lock<std::mutex>& lock)
+{
+    if (!lock.owns_lock())
+    {
+        lock.lock();
+    }
+
+    if (is_started && !spawner)
+    {
+        // If we should be running but a spawner does not exist, create one
+        spawner = std::make_unique<XWaylandSpawner>([this]() { spawn(); });
+    }
+}
+
+void mf::XWaylandConnector::clean_up(std::unique_lock<std::mutex>& lock)
+{
+    if (!lock.owns_lock())
+    {
+        lock.lock();
+    }
+
+    auto local_spawner{std::move(spawner)};
+    auto local_wm_event_thread{std::move(wm_event_thread)};
+    auto local_wm{std::move(wm)};
+    auto local_server{std::move(server)};
+
+    lock.unlock();
+
+    local_spawner.reset();
+    local_wm_event_thread.reset();
+    local_wm.reset();
+    local_server.reset();
```

Comment:
These area about to go out of scope, so resetting is redundant. (And *if*, for some dubious reason, the order matters, they can be defined in a suitable order.)

Reply:
I don't think the order matters, but I moved them around to keep it the same just in case. (current order has at least been tested)